File: Core/test/me/ihint/datahelper/test/java/impl/core/group/mysql/GroupTest.java
package me.ihint.datahelper.test.java.impl.core.group.mysql;

import me.ihint.datahelper.core.Data;
import me.ihint.datahelper.core.Field;
import me.ihint.datahelper.exception.bundle.FieldNotFoundException;
import me.ihint.datahelper.impl.core.bundle.SimpleBundle;
import me.ihint.datahelper.impl.core.config.SimpleConfig;
import me.ihint.datahelper.impl.core.datatype.mysql.INTEGER;
import me.ihint.datahelper.impl.core.datatype.mysql.VARCHAR;
import me.ihint.datahelper.impl.core.group.mysql.Record;
import me.ihint.datahelper.impl.core.group.mysql.Struct;

public class GroupTest {
    public static void test() {
        SimpleBundle<Field> fields = new SimpleBundle<>();
        fields.set("id", new Field("id", INTEGER.INSTANCE, new SimpleConfig()));

        SimpleConfig configOfUsername = new SimpleConfig();
        configOfUsername.set("length", 15);
        fields.set("username", new Field("username", VARCHAR.INSTANCE, configOfUsername));

        SimpleConfig configOfAge = new SimpleConfig();
        configOfAge.set("min", 0);
        fields.set("age", new Field("age", INTEGER.INSTANCE, configOfAge));

        SimpleConfig config = new SimpleConfig();
        config.set("writable", true);

        Struct struct = new Struct(fields, config);
        struct.setTableName("test_table_name");

        Record record = struct.newRecord();

        assert (record.getTableName().equals("test_table_name"));
        assert ((boolean) record.getConfig().get("writable") == true);

        Data id = record.getBundle().get("id");
        id.setValue("0");
        assert (record.get("id").equals("0"));

        record.set("username", "BlueSky");
        assert (record.get("username").equals("BlueSky"));

        record.getBundle().get("age").setValue("20");
        assert (record.get("age").equals("20"));

        try {
            record.get("null");
            throw new Exception();
        } catch (Exception e) {
            assert (e instanceof FieldNotFoundException);
        }

        try {
            record.set("null", "null");
            throw new Exception();
        } catch (Exception e) {
            assert (e instanceof FieldNotFoundException);
        }

        assert (record.count(false) == 3);
        assert (record.count(true) == 3);

        try {
            record.clear("null");
        } catch (Exception e) {
            assert (e instanceof FieldNotFoundException);
        }

        record.clear("username");
        assert (record.count(false) == 2);
        assert (record.count(true) == 3);

        record.clear();
        assert (record.count(false) == 0);
        assert (record.count(true) == 3);

        System.out.println("Group tests passed");
    }
}

File: Core/test/me/ihint/datahelper/test/kotlin/impl/core/group/mysql/GroupTest.kt
package me.ihint.datahelper.test.kotlin.impl.core.group.mysql

import me.ihint.datahelper.core.Field
import me.ihint.datahelper.exception.FieldNotFoundException
import me.ihint.datahelper.impl.core.bundle.SimpleBundle
import me.ihint.datahelper.impl.core.config.SimpleConfig
import me.ihint.datahelper.impl.core.datatype.mysql.INTEGER
import me.ihint.datahelper.impl.core.datatype.mysql.VARCHAR
import me.ihint.datahelper.impl.core.group.mysql.Struct

object GroupTest {
	fun test() {
		val fields = SimpleBundle<Field>()
		fields["id"] = Field("id", INTEGER, SimpleConfig())
		
		val configOfVarchar = SimpleConfig()
		configOfVarchar["length"] = 15
		fields["username"] = Field("id", VARCHAR, configOfVarchar)
		
		val configOfAge = SimpleConfig()
		configOfAge["min"] = 0
		fields["age"] = Field("age", INTEGER, configOfAge)
		
		val config = SimpleConfig()
		config.set("writable", true)
		
		val struct = Struct(fields, config)
		struct.setTableName("test_table_name")
		
		val record = struct.newRecord()
		
		assert(record.getTableName() == "test_table_name")
		assert(record.config["writable"] as Boolean == true)
		
		val id = record.bundle["id"] // first way to get data
		id!!.value = "0"
		assert(record.get("id") == "0") //  second way to get data
		
		record["username"] = "BlueSky" // third way to get data
		assert(record.get("username") == "BlueSky")
		
		record.bundle["age"]!!.value = "20"
		assert(record.get("age") == "20")
		
		try {
			record["null"]
			throw Exception()
		} catch (e: Exception) {
			assert(e is FieldNotFoundException)
		}
		
		try {
			record["null"] = "null"
			throw Exception()
		} catch (e: Exception) {
			assert(e is FieldNotFoundException)
		}
		
		println("Group tests passed")
	}
}
